Test to fail RBD mirroring when secondary gets full

sunilangadi2 opened the pull request:
# Description
Fixes https://issues.redhat.com/browse/RHCEPHQE-11620
Automated test case
https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-9507

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

manasagowri commented on suites/pacific/rbd/tier-3_rbd_mirror_secondary_full.yaml:
```diff
@@ -0,0 +1,175 @@
+      module: exec.py
+      name: configure mon_allow_pool_delete to True
+
+  - test:
```
What is the reason behind creating a separate test suite for this scenario alone?

sunilangadi2 replied:
It's `tier-3` scenario also it's OSD deployment is different than existing test suites hence to cover in weekly pipeline created separate test suite.

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_secondary_full.py:
```diff
@@ -0,0 +1,122 @@
+log = Log(__name__)
+
+
+def get_total_avail(cluster):
```
This could be a common utility so it can be moved to ceph/utils.py or cli/utilities/ folder.
Also please think of a better name which describes the method.

sunilangadi2 replied:
thanks i used the existing method from `utils.py`

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_secondary_full.py:
```diff
@@ -0,0 +1,122 @@
+        cluster2_size = get_total_avail(mirror2)
+
+        if cluster2_size > cluster1_size:
+            log.error("secondary cluster more than primary hence test cannot perform")
```
This need not be true. This will hold good only in the case when you are using primary mirroring alone to fill the secondary cluster. But secondary cluster could be filled by any other application running on it, or even by logs or any other reason. 

So this should not be a reason for you to not execute this test case.

sunilangadi2 replied:
added this validation thinking only `osd` storage space from primary and secondary cluster.

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_secondary_full.py:
```diff
@@ -0,0 +1,122 @@
+        log.debug(err)
+
+        # Validation for IO completion
+        if err:
```
Are you filling the cluster and also checking for mirroring in the same step? I don't think this is a good approach for this test case.

Afaik what you should do is, first fill the secondary cluster upto the stage where you start getting a HEALTH_ERR saying OSDs are full. 
Now try to create another image in primary and enable mirroring on this image, this should not succeed since ther e is no space left for the mirrored image to be created in secondary.

Instead you are just trying to write IOs to an already mirrored image which is not what this test case - CEPH-9507 is trying to test.

sunilangadi2 replied:
Done

manasagowri reviewed (changes requested):
I have provided my review for the partially complete implementation here as discussed. 
Please address these comments while completing the PR.

sunilangadi2 commented:
@manasagowri thanks for the review addressed all comments
success logs:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-ERYCED
please review it.

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_secondary_full.py:
```diff
@@ -0,0 +1,134 @@
+        # Validation for secondary full from secondary
+        out = mirror2.exec_cmd(cmd="ceph -s", output=True)
+
+        if "HEALTH_ERR" or "HEALTH_WARN" in out:
```
I think we discussed that the validation here should be that, it should be a HEALTH_ERR and also the secondary cluster health should say atleast one of the OSDs is full.

Please implement validation steps accordingly.

sunilangadi2 replied:
Done added

manasagowri commented on tests/rbd_mirror/test_rbd_mirror_secondary_full.py:
```diff
@@ -0,0 +1,134 @@
+
+        if "HEALTH_ERR" or "HEALTH_WARN" in out:
+            log.info(
+                "As Expected mirroring data gets failed, due to secondary cluster full"
```
Also, once secondary cluster is full, you should start adding a new image and enable mirroring on it, like I mentioned in my previous review as well and this should fail.
Please implement this as well.

sunilangadi2 replied:
@manasagowri can you please review this added new image mirroring after secondary gets full.

sunilangadi2 commented:
@manasagowri addressed all comments added verification steps with new image mirroring once secondary becomes full, buth there are multiple reasons the commands get failed so handled all in exceptions,
Test result:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-9EQE04
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-MF006V

sunilangadi2 commented:
@manasagowri As discussed implemented that way please review it.
Test results on newly added code
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-4RKMZ8

manasagowri reviewed (approved):
Changes LGTM
